process/process.go

```diff
@@ -730,6 +730,27 @@ func (p *Process) insertPropose(propose Propose) bool {
 		return false
 	}
 
+	// If we are here, it means the proposal is either valid, or was not passed
+	// through a Validator implementation. If its the latter case, we should
+	// not allow a nil value to be added to our propose logs, with the honest
+	// nodes trying to prevote and eventually precommit nil to advance round
+	if propose.Value == NilValue {
+		if p.broadcaster != nil {
+			p.broadcaster.BroadcastPrevote(Prevote{
+				Height: p.CurrentHeight,
+				Round:  propose.Round,
+				Value:  NilValue,
+				From:   p.whoami,
+			})
+		}
+
+		if p.CurrentRound == propose.Round {
+			p.stepToPrevoting()
+		}
+
+		return false
+	}
+
```

Comment:
```suggestion
```
I would prefer placing this as part of validation. That is, the moment we check validation, we explicitly check whether the value is `nil`.

Reply:
yes, we're already doing it (in Archaeon) and it HAS to be done for Darknode too. But I believe the audit report pointed it out because the `nil` value is not handled in Hyperdrive, and hence if the Hyperdrive user fails to handle it, `nil` values can be committed by accident.

Of course we should and will validate `nil` values. So should we scrap this change? In that case, I'll simply add a comment on `Valid` that hyperdrive expects the user to handle `nil` value proposals.